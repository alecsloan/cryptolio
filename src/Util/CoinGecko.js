export const getAssetData = async (currency, ids, assets) => {
  if (!currency) { return }

  try {
    return await window.fetch(`https://api.coingecko.com/api/v3/coins/markets?vs_currency=${currency}&ids=${ids}&price_change_percentage=1h%2C24h%2C7d`)
      .then(res => res.json())
      .then(response => {
        if (!response) { return }

        const newAssets = []

        Object.entries(response).forEach(responseAsset => {
          responseAsset = responseAsset[1]

          let cmcId = null
          let exitPlan = []
          let holdings = 0
          let interest = 0

          if (assets) {
            const existingAsset = assets.find(asset => asset.symbol === responseAsset.symbol.toUpperCase())

            if (existingAsset) {
              cmcId = existingAsset.cmcId
              exitPlan = existingAsset.exitPlan || exitPlan
              holdings = existingAsset.holdings || holdings
              interest = existingAsset.interest || interest
            }
          }

          newAssets.push({
            cgId: responseAsset.id,
            circulating_supply: responseAsset.circulating_supply,
            cmcId: cmcId,
            exitPlan: exitPlan,
            holdings: holdings,
            imageURL: responseAsset.image,
            interest: interest,
            market_cap: responseAsset.market_cap,
            max_supply: responseAsset.max_supply,
            name: responseAsset.name,
            percent_change_1h: responseAsset.price_change_percentage_1h_in_currency,
            percent_change_24h: responseAsset.price_change_percentage_24h_in_currency,
            percent_change_7d: responseAsset.price_change_percentage_7d_in_currency,
            price: responseAsset.current_price,
            rank: responseAsset.market_cap_rank,
            symbol: responseAsset.symbol.toUpperCase(),
            total_supply: responseAsset.total_supply,
            url: 'https://www.coingecko.com/en/coins/' + responseAsset.name.toLowerCase().replace(' ', '-'),
            volume_24h: responseAsset.total_volume
          })
        })

        return {
          assets: newAssets,
          timestamp: new Date()
        }
      })
  } catch (e) {
    console.log('Error getting CoinGecko data:', e)
  }
}

export const getAvailableAssets = async () => {
  try {
    return window.fetch('https://api.coingecko.com/api/v3/coins/list?include_platform=false')
      .then(res => res.json())
      .then(response => {
        const assets = []

        response.forEach(responseAsset => {
          assets.push({
            cgId: responseAsset.id,
            name: responseAsset.name,
            symbol: responseAsset.symbol
          })
        })

        return assets
      })
  } catch (e) {
    console.log('Error getting list of assets from CoinGecko. Check response.', e)
  }
}

export const getHistoricalAssetData = async (currency, ids, days = 1) => {
  if (!currency || !ids) { return }

  let time_end = (new Date().getTime() / 1000).toFixed(0);
  let time_start = time_end - (86400 * days);

    return Promise.all(
        ids.map((id) => {
          return window.fetch(`https://api.coingecko.com/api/v3/coins/${id}/market_chart/range?vs_currency=${currency}&to=${time_end}&from=${time_start}`)
        })
      )
      .then((responses) => {
        return Promise.all(
          responses.map((res) => {
            return res.json();
          }));
      })
      .then((response) => {
        return response
      })
      .catch((e) => {
        console.log('Error getting CoinGecko historical data:', e)
      })
}

export const logo = 'https://static.coingecko.com/s/coingecko-logo-d13d6bcceddbb003f146b33c2f7e8193d72b93bb343d38e392897c3df3e78bdd.png'
export const logoWhite = 'https://static.coingecko.com/s/coingecko-logo-white-3f2aeb48e13428b7199395259dbb96280bf47ea05b2940ef7d3e87c61e4d8408.png'
