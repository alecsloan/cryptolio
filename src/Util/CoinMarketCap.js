const CORS_PROXY = 'https://cors.bridged.cc/'
const GIRDA_API_KEY = '' //See https://github.com/girdaco/base/issues/23

export const getAssetData = async (currency, symbols, assets) => {
  if (!currency || !symbols || !GIRDA_API_KEY) { return }

  try {
    return await window.fetch(
        `${CORS_PROXY}https://web-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest?symbol=${symbols}&convert=${currency}`,
      {
          headers: {
            'x-cors-grida-api-key': `${GIRDA_API_KEY}`
          }
        }
      )
      .then(res => res.json())
      .then(response => {
        if (!response.data) { return }

        const newAssets = []

        Object.entries(response.data).forEach(responseAsset => {
          responseAsset = responseAsset[1]

          let cgId = null
          let cmcId = null
          let exitPlan = []
          let holdings = 0
          let interest = 0

          if (assets) {
            const existingAsset = assets.find(asset => asset.symbol === responseAsset.symbol)

            if (existingAsset) {
              cgId = existingAsset.cgId
              cmcId = existingAsset.cmcId
              exitPlan = existingAsset.exitPlan || exitPlan
              holdings = existingAsset.holdings || holdings
              interest = existingAsset.interest || interest
            }
          }

          if (!cmcId) { cmcId = responseAsset.id }

          newAssets.push({
            cgId: cgId,
            circulating_supply: responseAsset.circulating_supply,
            cmcId: cmcId,
            exitPlan: exitPlan,
            holdings: holdings,
            imageURL: 'https://s2.coinmarketcap.com/static/img/coins/128x128/' + cmcId + '.png',
            interest: interest,
            market_cap: responseAsset.quote[currency].market_cap,
            max_supply: responseAsset.max_supply,
            name: responseAsset.name,
            percent_change_1h: responseAsset.quote[currency].percent_change_1h,
            percent_change_24h: responseAsset.quote[currency].percent_change_24h,
            percent_change_7d: responseAsset.quote[currency].percent_change_7d,
            price: responseAsset.quote[currency].price,
            rank: responseAsset.cmc_rank,
            symbol: responseAsset.symbol,
            total_supply: responseAsset.total_supply,
            url: 'https://coinmarketcap.com/currencies/' + responseAsset.slug,
            volume_24h: responseAsset.quote[currency].volume_24h
          })
        })

        return {
          assets: newAssets,
          timestamp: response.status.timestamp
        }
      })
  } catch (e) {
    console.log('Error getting Coinmarketcap data:', e)
  }
}

export const getAvailableAssets = async () => {
  if (!GIRDA_API_KEY) return []

  try {
    return window.fetch(
        `${CORS_PROXY}https://web-api.coinmarketcap.com/v1/cryptocurrency/map`,
        {
          headers: {
            'x-cors-grida-api-key': `${GIRDA_API_KEY}`
          }
        }
      )
      .then(res => res.json())
      .then(response => {
        const assets = []

        response.data.forEach(responseAsset => {
          assets.push({
            cmcId: responseAsset.id,
            name: responseAsset.name,
            rank: responseAsset.rank,
            symbol: responseAsset.symbol
          })
        })

        return assets
      })
  } catch (e) {
    console.log('Error getting list of assets from CoinMarketCap. Check response.', e)
  }
}

export const getHistoricalAssetData = async (currency, symbols, days = 7) => {
  if (!currency || !symbols || !GIRDA_API_KEY) { return }

  let interval = 'hourly'

  if (days > 29) {
    interval = 'daily'
  }

  let time_end = (new Date().getTime() / 1000).toFixed(0)
  let time_start = time_end - Math.round(86400 * days)

  try {
    return await window.fetch(
        `${CORS_PROXY}https://web-api.coinmarketcap.com/v1/cryptocurrency/quotes/historical?format=chart_crypto_details&symbol=${symbols}&convert=${currency}&interval=${interval}&time_end=${time_end}&time_start=${time_start}`,
        {
          headers: {
            'x-cors-grida-api-key': `${GIRDA_API_KEY}`
          }
        }
      )
      .then(res => res.json())
      .then(response => {
        return response.data || null
      })
  } catch (e) {
    console.log('Error getting Coinmarketcap historical data:', e)
  }
}

const logoBase = 'https://s2.coinmarketcap.com/static/cloud/img/coinmarketcap'

export const logo = logoBase + '_1.svg'
export const logoWhite = logoBase + '_white_1.svg'
