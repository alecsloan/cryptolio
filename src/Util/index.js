import { Box, Typography } from '@material-ui/core'
import React from 'react'

export function getCurrencySymbol (currency) {
  return (0).toLocaleString(
    window.navigator.language,
    {
      style: 'currency',
      currency: currency,
      maximumFractionDigits: 0,
      minimumFractionDigits: 0
    }
  ).replace(/\d/g, '').trim()
}

export function getLocalizedNumber (number, settings) {
  return getLocalizedPrice(number, settings, 'decimal')
}

export function getLocalizedPrice (price, settings, style = 'currency') {
  if (!price || !settings) {
    return ''
  }

  let maxDigits = 0

  if (settings.decimals4 && price < settings.decimals4) {
    maxDigits = 4
  } else if (settings.decimals3 && price < settings.decimals3) {
    maxDigits = 3
  } else if (settings.decimals2 && price < settings.decimals2) {
    maxDigits = 2
  }

  return price.toLocaleString(
    window.navigator.language,
    {
      currency: settings.currency,
      maximumFractionDigits: maxDigits,
      minimumFractionDigits: maxDigits,
      style: style
    }
  )
}

export function getLocalizedPercent (number, minimumFractionDigits = 2) {
  return number.toLocaleString(window.navigator.language, { style: 'percent', minimumFractionDigits: minimumFractionDigits })
}

export function MyBalance (props) {
  if (props.holdings > 0 && props.settings.showAssetBalances) {
    return (
      <Typography component='div' id='balance'>
        <Box className='d-inline-block me-1' fontWeight='fontWeightBold'>
          Balance:
        </Box>
        <Box className='d-inline-block pl-2' fontWeight='fontWeightBold'>
          {getLocalizedPrice(props.price * props.holdings, props.settings)}
        </Box>
      </Typography>
    )
  }

  return (
    <div />
  )
}
