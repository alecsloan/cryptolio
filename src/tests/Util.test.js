import * as Util from '../Util/index'
import ReactDOM from 'react-dom'
import React from 'react'
import { MyBalance } from '../Util/index'

window.it('get currency symbol', () => {
  expect(Util.getCurrencySymbol('USD')).toBe('$')
  expect(Util.getCurrencySymbol('EUR')).toBe('€')
  expect(Util.getCurrencySymbol('GBP')).toBe('£')
})

window.it('get localized prices', () => {
  expect(Util.getLocalizedPrice(123456789.99, null)).toBe('')
  expect(Util.getLocalizedPrice(.12345, { currency: 'USD', decimals2: 1 })).toBe(`$0.12`)
  expect(Util.getLocalizedPrice(.12345, { currency: 'USD', decimals3: 1 })).toBe(`$0.123`)
  expect(Util.getLocalizedPrice(.12344, { currency: 'USD', decimals4: 1 })).toBe(`$0.1234`)
  expect(Util.getLocalizedPrice(.12345, { currency: 'USD', decimals4: 1 })).toBe(`$0.1235`)
  expect(Util.getLocalizedPrice(123456789.99, { currency: 'USD' })).toBe(`$123,456,790`)
  expect(Util.getLocalizedPrice(123456789.49, { currency: 'USD' })).toBe(`$123,456,789`)
})

window.it('get localized numbers', () => {
  expect(Util.getLocalizedNumber(123456789.99, null)).toBe('')
  expect(Util.getLocalizedNumber(.12345, { currency: 'USD', decimals2: 1 })).toBe(`0.12`)
  expect(Util.getLocalizedNumber(.12345, { currency: 'USD', decimals3: 1 })).toBe(`0.123`)
  expect(Util.getLocalizedNumber(.12344, { currency: 'USD', decimals4: 1 })).toBe(`0.1234`)
  expect(Util.getLocalizedNumber(.12345, { currency: 'USD', decimals4: 1 })).toBe(`0.1235`)
  expect(Util.getLocalizedNumber(123456789.99, { currency: 'USD' })).toBe(`123,456,790`)
  expect(Util.getLocalizedNumber(123456789.49, { currency: 'USD' })).toBe(`123,456,789`)
})

window.it('get localized percent', () => {
  expect(Util.getLocalizedPercent(.1234)).toBe('12.34%')
  expect(Util.getLocalizedPercent(.12345)).toBe('12.35%')
  expect(Util.getLocalizedPercent(.12345, 3)).toBe('12.345%')
  expect(Util.getLocalizedPercent(.12345, 1)).toBe('12.3%')
  expect(Util.getLocalizedPercent(.12345, 0)).toBe('12%')
  expect(Util.getLocalizedPercent(.125, 0)).toBe('13%')
})

window.it('MyBalance renders', () => {
  const holdings = 1
  const price = 123.50
  const settings = {
    currency: 'USD',
    showAssetBalances: true
  }
  
  const div = document.createElement('div')
  ReactDOM.render(<MyBalance holdings={holdings} price={price} settings={settings} />, div)

  expect(div).toMatchSnapshot()
  expect(div.textContent).toBe('Balance:$124')

  ReactDOM.unmountComponentAtNode(div)
})