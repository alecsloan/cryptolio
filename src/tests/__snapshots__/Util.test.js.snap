// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyBalance renders 1`] = `
<div>
  <div
    class="MuiTypography-root MuiTypography-body1"
    id="balance"
  >
    <div
      class="MuiBox-root MuiBox-root-1 d-inline-block me-1"
    >
      Balance:
    </div>
    <div
      class="MuiBox-root MuiBox-root-2 d-inline-block pl-2"
    >
      $124
    </div>
  </div>
</div>
`;
